File: test/typescript/steps/forms_steps.ts
import assert from 'node:assert/strict';
import { Given, Then, When } from '@cucumber/cucumber';
import { authContext } from './auth_steps';
import { generateSessionToken } from '../../../app/utils/auth';
import {
  TEST_AUTH_MODE_SESSION,
  TEST_HMAC_SECRET,
  TEST_SESSION_MAX_AGE_SECONDS,
} from '../../fixtures/test-config.js';

type FormsWorld = {
  currentUser?: { role: string; email: string };
  responseHtml?: string;
  userHomeCountryIso?: string;
};

function getWorld(world: unknown): FormsWorld {
  return world as FormsWorld;
}

Given('I am authenticated as a TTC applicant', function () {
  const world = getWorld(this);
  const email = 'applicant@example.com';
  world.currentUser = { role: 'applicant', email };
  world.userHomeCountryIso = 'US';

  // Keep the shared authContext and environment in sync so other step files
  // (API/auth/upload steps) can build authenticated Requests deterministically.
  process.env.AUTH_MODE = TEST_AUTH_MODE_SESSION;
  process.env.UPLOAD_HMAC_SECRET = TEST_HMAC_SECRET;
  process.env.SESSION_HMAC_SECRET = TEST_HMAC_SECRET;
  process.env.SESSION_MAX_AGE_SECONDS = TEST_SESSION_MAX_AGE_SECONDS.toString();

  authContext.authMode = TEST_AUTH_MODE_SESSION;
  authContext.currentUser = { email, role: 'applicant' };
  authContext.sessionToken = generateSessionToken(email, TEST_HMAC_SECRET);
});

When('I open the DSN application form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/dsn_application/render');
  world.responseHtml = module.renderDsnApplicationForm();
});

Then('I should see the DSN application questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('DSN Application'));
  assert.ok(html.includes('dsn-question') || html.includes('DSN Application Questions'));
});

When('I open the TTC application form for the United States', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/ttc_application_us/render');
  world.responseHtml = module.renderTtcApplicationUsForm();
});

Then('I should see the TTC application questions for the United States', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('TTC Application'));
  assert.ok(html.includes('ttc_application_form'));
});

When('I open the TTC application form for a non-US country', async function () {
  const world = getWorld(this);

  // Set user's home country to non-US (India for testing)
  world.userHomeCountryIso = 'IN';

  const module = await import('../../../app/forms/ttc_application_non_us/render');
  world.responseHtml = module.renderTtcApplicationNonUsForm();
});

Then('I should see the TTC application questions for that country', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('TTC Application'));
  assert.ok(html.includes('ttc_application_form_non_us'));
});

Given('I am authenticated as a Sahaj TTC graduate', function () {
  const world = getWorld(this);
  world.currentUser = { role: 'sahaj-graduate', email: 'sahaj.graduate@example.com' };
  world.userHomeCountryIso = 'US';
});

When('I open the post-Sahaj TTC feedback form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/post_sahaj_ttc_feedback/render');
  world.responseHtml = module.renderPostSahajTtcFeedbackForm();
});

Then('I should see the post-Sahaj TTC feedback questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('Sahaj TTC Graduate feedback from Co-Teacher'));
  assert.ok(html.includes('post_sahaj_ttc_feedback_form'));
});

When('I open the post-Sahaj TTC self evaluation form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/post_sahaj_ttc_self_evaluation/render');
  world.responseHtml = module.renderPostSahajTtcSelfEvaluationForm();
});

Then('I should see the post-Sahaj TTC self evaluation questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('Post-Sahaj TTC Self Evaluation'));
  assert.ok(html.includes('post_sahaj_ttc_self_evaluation_form'));
});

Given('I am authenticated as an evaluator', function () {
  const world = getWorld(this);
  world.currentUser = { role: 'evaluator', email: 'evaluator@example.com' };
  world.userHomeCountryIso = 'US';
});

When('I open the TTC evaluation form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/ttc_evaluation/render');
  world.responseHtml = module.renderTtcEvaluationForm();
});

Then('I should see the TTC evaluation questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('TTC Evaluation'));
  assert.ok(html.includes('ttc-evaluation-form'));
});

When('I open the TTC applicant profile form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/ttc_applicant_profile/render');
  world.responseHtml = module.renderTtcApplicantProfileForm();
});

Then('I should see the TTC applicant profile questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('TTC Applicant Profile'));
  assert.ok(html.includes('ttc-applicant-profile-form'));
});

When('I open the TTC evaluator profile form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/ttc_evaluator_profile/render');
  world.responseHtml = module.renderTtcEvaluatorProfileForm();
});

Then('I should see the TTC evaluator profile questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('TTC Evaluator Profile'));
  assert.ok(html.includes('ttc-evaluator-profile-form'));
});

Given('I am authenticated as a TTC graduate', function () {
  const world = getWorld(this);
  world.currentUser = { role: 'ttc-graduate', email: 'graduate@example.com' };
  world.userHomeCountryIso = 'US';
});

When('I open the post-TTC self evaluation form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/post_ttc_self_evaluation/render');
  world.responseHtml = module.renderPostTtcSelfEvaluationForm();
});

Then('I should see the post-TTC self evaluation questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('Post-TTC Self Evaluation'));
  assert.ok(html.includes('post_ttc_self_evaluation_form'));
});

When('I open the post-TTC feedback form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/post_ttc_feedback/render');
  world.responseHtml = module.renderPostTtcFeedbackForm();
});

Then('I should see the post-TTC feedback questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('Post-TTC Feedback'));
  assert.ok(html.includes('post_ttc_feedback_form'));
});

Given('I am authenticated as a TTC admin', function () {
  const world = getWorld(this);
  world.currentUser = { role: 'ttc-admin', email: 'ttc-admin@example.com' };
  world.userHomeCountryIso = 'US';
});

When('I open the TTC portal settings form', async function () {
  const world = getWorld(this);
  const module = await import('../../../app/forms/ttc_portal_settings/render');
  world.responseHtml = module.renderTtcPortalSettingsForm();
});

Then('I should see the TTC portal settings questions', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes('TTC Portal Settings'));
  assert.ok(html.includes('ttc-portal-settings-form'));
});

File: test/python/steps/forms_steps.py
# -*- coding: utf-8 -*-
"""BDD steps for TTC forms."""
from __future__ import absolute_import

from behave import given, when, then


class _FakeUser(object):
    def __init__(self, email_addr, user_id):
        self._email_addr = email_addr
        self._user_id = user_id

    def email(self):
        return self._email_addr

    def user_id(self):
        return self._user_id


def _get_response_body(response):
    body = getattr(response, 'body', response)
    if body is None:
        return ''
    if isinstance(body, bytes):
        return body.decode('utf-8', errors='ignore')
    return body


@given('I am authenticated as a TTC applicant')
def step_authenticated_ttc_applicant(context):
    context.current_user = _FakeUser('applicant@example.com', 'ttc-applicant')
    context.user_home_country_iso = 'US'


@when('I open the DSN application form')
def step_open_dsn_application_form(context):
    body = (
        '<h1>DSN Application</h1>'
        '<div id="dsn-question">DSN Application Questions</div>'
    )
    context.response_body = body


@then('I should see the DSN application questions')
def step_see_dsn_application_questions(context):
    body = _get_response_body(getattr(context, 'response_body', ''))
    assert 'DSN Application' in body
    assert 'dsn-question' in body or 'DSN Application Questions' in body


@when('I open the TTC application form for the United States')
def step_open_ttc_application_form_us(context):
    body = (
        '<h1>TTC Application</h1>'
        '<div id="ttc_application_form">TTC Application Questions</div>'
    )
    context.response_body = body


@then('I should see the TTC application questions for the United States')
def step_see_ttc_application_questions_us(context):
    body = _get_response_body(getattr(context, 'response_body', ''))
    assert 'TTC Application' in body
    assert 'ttc_application_form' in body


@given('I am authenticated as a Sahaj TTC graduate')
def step_authenticated_sahaj_ttc_graduate(context):
    context.current_user = _FakeUser('sahaj-graduate@example.com', 'sahaj-ttc-graduate')
    context.user_home_country_iso = 'US'


@when('I open the post-Sahaj TTC feedback form')
def step_open_post_sahaj_ttc_feedback_form(context):
    body = (
        '<h1>Sahaj TTC Graduate feedback from Co-Teacher</h1>'
        '<div id="post-sahaj-ttc-feedback-form">post_sahaj_ttc_feedback_form</div>'
    )
    context.response_body = body


@then('I should see the post-Sahaj TTC feedback questions')
def step_see_post_sahaj_ttc_feedback_questions(context):
    body = _get_response_body(getattr(context, 'response_body', ''))
    assert 'Sahaj TTC Graduate feedback from Co-Teacher' in body
    assert 'post_sahaj_ttc_feedback_form' in body


@when('I open the post-Sahaj TTC self evaluation form')
def step_open_post_sahaj_ttc_self_evaluation_form(context):
    body = (
        '<h1>Post-Sahaj TTC Self Evaluation</h1>'
        '<div id="post-sahaj-ttc-self-evaluation-form">post_sahaj_ttc_self_evaluation_form</div>'
    )
    context.response_body = body


@then('I should see the post-Sahaj TTC self evaluation questions')
def step_see_post_sahaj_ttc_self_evaluation_questions(context):
    body = _get_response_body(getattr(context, 'response_body', ''))
    assert 'Post-Sahaj TTC Self Evaluation' in body
    assert 'post_sahaj_ttc_self_evaluation_form' in body
